### PR TITLE
Update std::shared_ptr argument passing to be closer to idiomatic C++.

### PR DESCRIPTION
**Pull request by jpieper-tri**
The CppCoreGuidelines recommend that std::shared_ptr only be passed as
an argument when the callee explicitly needs to manage the lifetime of
the argument, and that 'const std::shared_ptr<T>&' should only be used
if there is a chance that lifetime will not be shared.

https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rr-smartptrparam

Many instances in drake were passing 'const std::shared_ptr<T>&'
around, often when no lifetime management was necessary, and otherwise
when lifetime sharing was unconditional.  This can be less efficient,
unnecessarily restrictive, and hides the intent of the API.

This patch converts those instances either to:
 * [const] T& if no lifetime sharing is required
 * std::shared_ptr<T> if lifetime sharing is always required

### CODE REVIEW

**Comment by sherm1** [comment]
Some build failures, e.g.:
```
/tmp/drake-1797-9134/drake/systems/plants/constructModelmex.cpp:254:45: error: non-const lvalue reference to type 'RigidBody' cannot bind to a value of unrelated type 'shared_ptr<RigidBody>'
        model->addCollisionElement(element, b, group_name);
                                            ^
/tmp/drake-1797-9134/drake/../drake/systems/plants/RigidBodyTree.h:526:18: note: passing argument to parameter 'body' here
      RigidBody& body, const std::string& group_name);
                 ^
```

**Comment by jpieper-tri** [comment]
sherm1: Thanks, hopefully fixed, although I'll have to rely on CI as I don't have matlab locally to test.

File: drake/systems/plants/RigidBody.h [hunk 1]
```diff
@@ -56,17 +56,17 @@ class DRAKERBM_EXPORT RigidBody {
     this->collision_filter_ignores &= ~group;
   };
 
-  bool adjacentTo(const std::shared_ptr<RigidBody>& other) const {
-    return ((parent == other && !(joint && joint->isFloating())) ||
-            (other->parent.get() == this &&
-             !(other->joint && other->joint->isFloating())));
+  bool adjacentTo(const RigidBody& other) const {
+    return ((parent.get() == &other && !(joint && joint->isFloating())) ||
```
Comment (sherm1) [inline drake/systems/plants/RigidBody.h]: This and some of the nearby methods would be clearer with a utility method something like
```c++
return (parent->isSameRigidBody(other) && ...);
```

**Comment by sherm1** [comment]
Looks good to me, pending builds.

**Comment by jpieper-tri** [comment]
Well, the windows jenkins build failed, and while the build output is full of warnings, I think the actual failure was just #1787 ?

**Comment by sherm1** [comment]
> the windows jenkins build failed ... I think the actual failure was just #1787?

Yeah, the `patch` command is missing.

**Comment by jpieper-tri** [comment]
Updated to fix merge conflicts.

**Comment by jpieper-tri** [comment]
Does anyone know why jenkins failed here?  It looks like it was trying to pull the incorrect git hash?

**Comment by jhoare** [comment]
retest this please

**Comment by jpieper-tri** [comment]
Great, I believe the spurious Windows jenkins failures are now resolved.

**Comment by RussTedrake** [comment]
@jamiesnape -- why did this pass?  it should not have passed without #1843 being merged.

**Comment by jamiesnape** [comment]
@billhoffman added an environment variable for now.

**Comment by jamiesnape** [comment]
(in the Jenkins build script, that is)

**Comment by RussTedrake** [comment]
that doesn't explain why the pendulum urdf test passed.

**Pull request merged**